codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ManifestWriter.java

```diff
@@ -0,0 +1,157 @@
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import software.amazon.smithy.build.FileManifest;
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.codegen.core.SymbolDependency;
+import software.amazon.smithy.model.node.ArrayNode;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.StringNode;
+
+/**
+ * Generates a manifest description of the generated code, minimum go version, and minimum dependencies required.
+ */
+public final class ManifestWriter {
+    public static final String INCOMPATIBLE_VERSION = "+incompatible";
+
+    private ManifestWriter() {
+    }
+
+    /**
+     * Write the manifest description of the generated code.
+     *
+     * @param settings the go settings
+     * @param fileManifest the file manifest
+     * @param dependencies the list of symbol dependencies
+     */
+    public static void writeManifest(
+            GoSettings settings,
+            FileManifest fileManifest,
+            List<SymbolDependency> dependencies
+    ) {
+        Path manifestFile = fileManifest.getBaseDir().resolve("generated.json");
+
+        if (Files.exists(manifestFile)) {
+            try {
+                Files.delete(manifestFile);
+            } catch (IOException e) {
+                throw new CodegenException("Failed to delete existing generated.json file", e);
+            }
+        }
+        fileManifest.addFile(manifestFile);
+
+        ObjectNode generatedJson = buildManifestFile(settings, fileManifest, dependencies);
+        fileManifest.writeFile(manifestFile.toString(), Node.prettyPrintJson(generatedJson));
+    }
+
+    private static ObjectNode buildManifestFile(
+            GoSettings settings,
+            FileManifest fileManifest,
+            List<SymbolDependency> dependencies
+    ) {
+        Stream<SymbolDependency> nonStdLib = dependencies.stream()
+                .filter(symbolDependency ->
+                        !symbolDependency.getDependencyType().equals(GoDependency.Type.STANDARD_LIBRARY.toString()));
+
+        Optional<SymbolDependency> minStandard = dependencies.stream()
+                .filter(symbolDependency ->
+                        symbolDependency.getDependencyType().equals(GoDependency.Type.STANDARD_LIBRARY.toString()))
+                .max(Comparator.comparing(SymbolDependency::getVersion));
+
+        Map<StringNode, Node> manifestNodes = new HashMap<>();
+
+        Map<String, String> minimumDependencies = gatherMinimumDependencies(nonStdLib);
+
+        Map<StringNode, Node> dependencyNodes = new HashMap<>();
```

Comment:
Should we use TreeMap's here to generate a stable list?

Reply:
Assuming this file would be checked in as a part of the repo, its content would need to be stable.